src/coprocessor/dag.rs

```diff
@@ -0,0 +1,147 @@
+               deadline: Instant,
+               ranges: Vec<KeyRange>,
+               snap: &'s Snapshot,
+               eval_ctx: Rc<EvalContext>)
+               -> DAGContext<'s> {
+        DAGContext {
+            req: req,
+            deadline: deadline,
+            columns: vec![],
+            ranges: ranges,
+            snap: snap,
+            has_aggr: false,
+            eval_ctx: eval_ctx,
+            chunks: vec![],
+        }
+    }
+
+    pub fn validate_dag(&mut self) -> Result<()> {
+        let execs = self.req.get_executors();
+        let first = try!(execs.first()
+            .ok_or_else(|| Error::Other(box_err!("dag copr has not executor"))));
+        // check whether first exec is *scan and get the column info
+        match first.get_tp() {
+            ExecType::TypeTableScan => {
+                self.columns = first.get_tbl_scan().get_columns().to_vec();
+            }
+            ExecType::TypeIndexScan => {
+                self.columns = first.get_idx_scan().get_columns().to_vec();
+            }
+            _ => {
+                return Err(box_err!("first exec type should be *Scan, but get {:?}",
+                                    first.get_tp()))
+            }
+        }
+        // check whether dag has a aggregation action and take a flag
+        if execs.iter().rev().any(|ref exec| exec.get_tp() == ExecType::TypeAggregation) {
+            self.has_aggr = true;
+        }
+        Ok(())
+    }
+
+    // FIXME seperate first exec build action from `build_dag`
+    // since it will generte some mutable conflict when putting together
+    pub fn build_first(&'s self,
+                       mut first: Executor,
+                       statistics: &'s mut Statistics)
+                       -> Box<DAGExecutor + 's> {
+        match first.get_tp() {
+            ExecType::TypeTableScan => {
+                Box::new(TableScanExecutor::new(first.take_tbl_scan(),
+                                                self.ranges.clone(),
+                                                self.snap,
+                                                statistics,
+                                                self.req.get_start_ts(),
+                                                IsolationLevel::SI))
+            }
+            ExecType::TypeIndexScan => {
+                Box::new(IndexScanExecutor::new(first.take_idx_scan(),
+                                                self.ranges.clone(),
+                                                self.snap,
+                                                statistics,
+                                                self.req.get_start_ts(),
+                                                IsolationLevel::SI))
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn build_dag(&'s self, statistics: &'s mut Statistics) -> Result<Box<DAGExecutor + 's>> {
+        let mut execs = self.req.get_executors().to_vec().into_iter();
```

Comment:
use reference here will generate a mut borrow conflict in the env which used this func.